utils/src/main/kotlin/OrtProxySelector.kt

```diff
@@ -0,0 +1,211 @@
+ */
+class OrtProxySelector(private val fallback: ProxySelector? = null) : ProxySelector() {
+    companion object {
+        private val NO_PROXY_LIST = listOf(Proxy.NO_PROXY)
+
+        /**
+         * Install this proxy selector as the global default. The previous selector is used as a fallback.
+         */
+        fun install(): OrtProxySelector {
+            val current = getDefault()
+            return if (current is OrtProxySelector) {
+                log.info { "Proxy selector is already installed." }
+                current
+            } else {
+                OrtProxySelector(current).also {
+                    setDefault(it)
+                    log.info { "Proxy selector was successfully installed." }
+                }
+            }
+        }
+
+        /**
+         * Uninstall this proxy selector, restoring the previous selector as the global default.
+         */
+        fun uninstall(): ProxySelector? {
+            val current = getDefault()
+            return if (current is OrtProxySelector) {
+                current.fallback.also {
+                    setDefault(it)
+                    log.info { "Proxy selector was successfully uninstalled." }
+                }
+            } else {
+                log.info { "Proxy selector is not installed." }
+                current
+            }
+        }
+    }
+
+    private val proxyAuthentication = mutableMapOf<Proxy, PasswordAuthentication?>()
+    private val proxyOrigins = mutableMapOf<String, MutableMap<String, MutableList<Proxy>>>()
+
+    init {
+        determineProxyFromProperties("http")?.let {
+            addProxy("properties", "http", it)
+        }
+
+        determineProxyFromProperties("https")?.let {
+            addProxy("properties", "https", it)
+        }
+
+        determineProxyFromURL(Os.env["http_proxy"])?.let {
+            addProxy("environment", "http", it)
+        }
+
+        determineProxyFromURL(Os.env["https_proxy"])?.let {
+            addProxy("environment", "https", it)
+        }
+    }
+
+    /**
+     * Add a [proxy with optional password authentication][authenticatedProxy] that can handle the [protocol] and whose
+     * definition comes from the [origin].
+     */
+    fun addProxy(origin: String, protocol: String, authenticatedProxy: AuthenticatedProxy) {
+        val (proxy, authentication) = authenticatedProxy
+        proxyAuthentication[proxy] = authentication
+
+        val proxiesForOrigin = proxyOrigins.getOrPut(origin) { mutableMapOf() }
+        val proxiesForProtocol = proxiesForOrigin.getOrPut(protocol) { mutableListOf() }
+        proxiesForProtocol += proxy
+    }
+
+    /**
+     * Return whether the [proxy] has been defined from some origin.
+     */
+    fun hasOrigin(proxy: Proxy) =
+        proxy in proxyOrigins.flatMap { (_, proxiesForProtocol) ->
+            proxiesForProtocol.values
+        }.flatten()
```

Comment:
Is `flatten` still required although `flatMap` is used above?

Reply:
Yes. `flatMap()` here only flattens one nesting level.